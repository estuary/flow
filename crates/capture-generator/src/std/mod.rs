use async_trait::async_trait;
use http::header::AUTHORIZATION;
use reqwest::{Request, Url};
use schemars::{schema::RootSchema, JsonSchema};
use serde::Deserialize;

use crate::interface::*;

#[derive(Default)]
pub struct BearerTokenAuth {
    config: Option<BearerTokenConfig>,
}
#[derive(Deserialize, JsonSchema)]
pub struct BearerTokenConfig {
    token: String,
}
#[async_trait]
impl Auth for BearerTokenAuth {
    type Config = BearerTokenConfig;

    fn prepare(&mut self, config: BearerTokenConfig) -> Result<(), anyhow::Error> {
        self.config = Some(config);
        Ok(())
    }

    async fn authenticate(&mut self, req: &mut Request) -> Result<(), anyhow::Error> {
        let token = &self.config.as_ref().unwrap().token;
        let headers = req.headers_mut();
        headers.insert(AUTHORIZATION, format!("Bearer {token}").parse().unwrap());
        Ok(())
    }
}

/// A very basic stream that sends a HTTP request to a specific URL.
/// Can be used alongside different paginations to create a stream.
pub struct BasicStream {
    pub key: String,
    pub description: String,
    pub endpoint: String,
    pub method: http::Method,
    pub spec: RootSchema,
}
impl Stream for BasicStream {
    type Config = ();

    fn key(&mut self) -> String {
        self.key.clone()
    }

    fn description(&mut self) -> String {
        self.description.clone()
    }

    fn create_request(&mut self) -> Request {
        let mut request = Request::new(self.method.clone(), Url::parse(&self.endpoint).unwrap());
        request
            .headers_mut()
            .insert("User-Agent", "flow-capture".parse().unwrap());

        let body = request.body_mut();
        *body = Some(vec![].into());

        request
    }

    fn spec(&mut self) -> RootSchema {
        self.spec.clone()
    }
}

pub struct BasicPagination {
    pub page_query_field: String,
    pub current_page_field: String,
    pub max_pages_field: String,
}

#[async_trait]
impl Pagination for BasicPagination {
    type Config = ();

    async fn paginate(
        &mut self,
        req: &mut reqwest::Request,
        last_response: reqwest::Response,
    ) -> Result<bool, anyhow::Error> {
        let last_body = last_response.text().await?;
        println!("last_body {:#?}", last_body);
        let last_body = serde_json::from_str::<serde_json::Value>(&last_body)?;

        let res_current_page = last_body.get(&self.current_page_field).unwrap();
        let res_max_page = last_body.get(&self.max_pages_field).unwrap();

        if res_current_page == res_max_page {
            Ok(false)
        } else {
            let current_page = res_current_page.as_u64().unwrap();
            let next_page = (current_page + 1).to_string();

            req.url_mut()
                .query_pairs_mut()
                .append_pair(&self.page_query_field, &next_page);

            println!("{:#?}", req.url().query());

            Ok(true)
        }
    }
}
