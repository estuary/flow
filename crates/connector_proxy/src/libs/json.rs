use doc::ptr::{Pointer, Token};
use schemars::{schema::RootSchema, JsonSchema};
use serde_json::Value;

// Create the RootSchema given datatype T.
pub fn create_root_schema<T: JsonSchema>() -> RootSchema {
    let mut settings = schemars::gen::SchemaSettings::draft07();
    settings.inline_subschemas = true;
    let generator = schemars::gen::SchemaGenerator::new(settings);
    return generator.into_root_schema_for::<T>();
}

// Extract the sub object keyed at `key` from v if v is an object, and returns both the remainder and the removed value.
pub fn remove_subobject(mut v: Value, key: &str) -> (Option<Value>, Value) {
    let mut sub_object = None;

    if let Value::Object(ref mut m) = v {
        sub_object = m.remove(key)
    }

    (sub_object, v)
}

pub fn tokenize_jsonpointer(ptr: &str) -> Vec<String> {
    Pointer::from_str(&ptr)
        .iter()
        .map(|t| match t {
            // Keep the index and next index for now. Could adjust based on usecases.
            Token::Index(ind) => ind.to_string(),
            Token::Property(prop) => prop.to_string(),
            Token::NextIndex => "-".to_string(),
        })
        .collect()
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_tokenize_jsonpointer() {
        let expected: Vec<String> = vec!["p1", "p2", "56", "p3", "-"]
            .iter()
            .map(|s| s.to_string())
            .collect();

        assert!(expected == tokenize_jsonpointer("/p1/p2/56/p3/-"));
    }
}
