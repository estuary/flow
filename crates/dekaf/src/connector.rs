use anyhow::{bail, Context};
use proto_flow::{flow::materialization_spec, materialize};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema, Copy)]
#[serde(rename_all = "snake_case")]
pub enum DeletionMode {
    // Handles deletions using the regular Kafka upsert envelope, where a deletion
    // is represented by a record containing the key that was deleted, and a null value.
    Kafka,
    // Handles deletions by passing through the full deletion document as it exists
    // in the source collection, as well as including a new field `_meta/is_deleted`
    // which is defined as the number `1` on deletions, and `0` otherwise.
    #[serde(rename = "cdc")]
    CDC,
}

impl Default for DeletionMode {
    fn default() -> Self {
        Self::Kafka
    }
}

/// Configures the behavior of a whole dekaf task
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
pub struct DekafConfig {
    /// Whether or not to expose topic names in a strictly Kafka-compliant format
    /// for systems that require it. Off by default.
    pub strict_topic_names: bool,
    /// The password that will authenticate Kafka consumers to this task.
    // TODO(jshearer): Uncomment when schemars 1.0 is out and we upgrade
    // #[schemars(extend("secret" = true))]
    #[schemars(schema_with = "token_secret")]
    pub token: String,
    /// How to handle deletion events. "Default" emits them as regular Kafka
    /// tombstones with null values, and "Header" emits then as a kafka document
    /// with empty string and `_is_deleted` header set to `1`. Setting this value
    /// will also cause all other non-deletions to have an `_is_deleted` header of `0`.
    #[serde(default)]
    pub deletions: DeletionMode,
}

/// Configures a particular binding in a Dekaf-type materialization
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
pub struct DekafResourceConfig {
    /// The exposed name of the topic that maps to this binding. This
    /// will be exposed through the Kafka metadata/discovery APIs.
    #[schemars(schema_with = "collection_name")]
    pub topic_name: String,
}

fn collection_name(_gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
    serde_json::from_value(serde_json::json!({
        "x-collection-name": true,
    }))
    .unwrap()
}

fn token_secret(_gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
    serde_json::from_value(serde_json::json!({
        "title": "Dekaf Auth Token",
        "secret": true,
    }))
    .unwrap()
}

pub async fn unary_materialize(
    request: materialize::Request,
) -> anyhow::Result<materialize::Response> {
    use proto_flow::materialize::response::validated;
    if let Some(mut validate) = request.validate {
        match materialization_spec::ConnectorType::try_from(validate.connector_type)? {
            materialization_spec::ConnectorType::Dekaf => {}
            other => bail!("invalid connector type: {}", other.as_str_name()),
        };

        let parsed_outer_config =
            serde_json::from_str::<models::DekafConfig>(&validate.config_json)
                .context("validating dekaf config")?;

        let _parsed_inner_config = serde_json::from_value::<DekafConfig>(
            parsed_outer_config.config.to_value(),
        )
        .context(format!(
            "validating dekaf endpoint config for variant {}",
            parsed_outer_config.variant
        ))?;

        // Largely copied from crates/validation/src/noop.rs
        let validated_bindings = std::mem::take(&mut validate.bindings)
            .into_iter()
            .enumerate()
            .map(|(i, b)| {
                let resource_path = vec![format!("binding-{}", i)];
                let constraints = b
                    .collection
                    .expect("collection must exist")
                    .projections
                    .into_iter()
                    .map(|proj| {
                        (
                            proj.field,
                            validated::Constraint {
                                r#type: validated::constraint::Type::FieldOptional as i32,
                                reason: "Dekaf allows everything for now".to_string(),
                            },
                        )
                    })
                    .collect::<BTreeMap<_, _>>();
                validated::Binding {
                    constraints,
                    resource_path,
                    delta_updates: false,
                }
            })
            .collect::<Vec<_>>();

        return Ok(materialize::Response {
            validated: Some(materialize::response::Validated {
                bindings: validated_bindings,
            }),
            ..Default::default()
        });
    } else {
        bail!("Unhandled request type")
    }
}
