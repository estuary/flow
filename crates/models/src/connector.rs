use super::RawValue;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

/// Splits a full connector image name into separate image and tag components.
/// The resulting tag will always begin with either a `@sha256:` or `:` if a
/// tag is present. Otherwise, the tag will be an empty string.
pub fn split_image_tag(image_full: &str) -> (String, String) {
    let mut image = image_full.to_string();

    if let Some(pivot) = image.find("@sha256:").or_else(|| image.find(":")) {
        let tag = image.split_off(pivot);
        (image, tag)
    } else {
        (image, String::new())
    }
}

/// Dekaf service configuration
#[derive(Serialize, Deserialize, Clone, Debug, JsonSchema, PartialEq)]
pub struct DekafConfig {
    /// # Dekaf variant type. Since we support integrating with a bunch of different providers via Dekaf,
    /// this allows us to store which of those connector variants this particular Dekaf connector was
    /// created as, in order to e.g link to the correct docs URL, show the correct name and logo, etc.
    pub variant: String,
    /// # Dekaf endpoint config.
    pub config: RawValue,
}

/// Connector image and configuration specification.
#[derive(Serialize, Deserialize, Clone, Debug, JsonSchema, PartialEq)]
pub struct ConnectorConfig {
    /// # Image of the connector.
    pub image: String,
    /// # Configuration of the connector.
    pub config: RawValue,
}

impl ConnectorConfig {
    pub fn example() -> Self {
        Self {
            image: "connector/image:tag".to_string(),
            config: serde_json::from_str("\"connector-config.yaml\"").unwrap(),
        }
    }
}

/// Local command and its configuration.
#[derive(Serialize, Deserialize, Clone, Debug, JsonSchema, PartialEq)]
pub struct LocalConfig {
    /// # Command to execute
    pub command: Vec<String>,
    /// # Configuration of the command.
    pub config: RawValue,
    /// # Environment variables
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub env: BTreeMap<String, String>,
    /// # Use protobuf codec instead of JSON.
    #[serde(default, skip_serializing_if = "super::is_false")]
    pub protobuf: bool,
}

impl LocalConfig {
    pub fn example() -> Self {
        Self {
            command: vec![
                "my-connector".to_string(),
                "--arg=one".to_string(),
                "--arg=two".to_string(),
            ],
            config: serde_json::from_value(serde_json::json!({"field": "value", "otherField": 42}))
                .unwrap(),
            env: BTreeMap::new(),
            protobuf: false,
        }
    }
}
