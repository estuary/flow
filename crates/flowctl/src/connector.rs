use anyhow::anyhow;
use std::{fs, process::{Command, Stdio, Output, Child}, pin::Pin};
use anyhow::Context;
use futures::{stream, TryStream, Stream};
use proto_flow::capture::{Request, Response};
use tempfile::{tempdir, TempDir};
use proto_grpc::capture::connector_client::ConnectorClient;

fn pull(image: &str) -> anyhow::Result<Output> {
    Command::new("docker")
        .args(["pull", image])
        .output().map_err(|e| e.into())
}

fn inspect(image: &str) -> anyhow::Result<Output> {
    Command::new("docker")
        .args(["inspect", image])
        .output().map_err(|e| e.into())
}

const CONNECTOR_INIT_PORT: u16 = 49092;

pub fn docker_spawn(image: &str, args: &[&str]) -> anyhow::Result<(Child, TempDir, u16)> {
    pull(image).context(format!("pulling {image}"))?;

    let inspect_output = inspect(image).context(format!("inspecting {image}"))?;

    let target_inspect = "/tmp/image-inspect.json";
    let dir = tempdir().context("creating temp directory")?;
    let host_inspect = dir.path().join("image-inspect.json");
    let host_inspect_str = host_inspect.clone().into_os_string().into_string().unwrap();

    fs::write(&host_inspect, inspect_output.stdout)?;
    let host_connector_init = locate_bin::locate("flow-connector-init").context("locating flow-connector-init")?;
    let host_connector_init_str = host_connector_init.into_os_string().into_string().unwrap();
    let target_connector_init = "/tmp/connector_init";

    let port = portpicker::pick_unused_port().expect("No ports free");

    let child = Command::new("docker")
        .args([&[
              "run",
              "--interactive",
              "--init",
              "--rm",
              "--log-driver=none",
              "--mount",
              &format!("type=bind,source={host_inspect_str},target={target_inspect}"),
              "--mount",
              &format!("type=bind,source={host_connector_init_str},target={target_connector_init}"),
		      "--publish", &format!("0.0.0.0:{}:{}/tcp", port, CONNECTOR_INIT_PORT),
              "--entrypoint",
              target_connector_init,
              image,
              &format!("--image-inspect-json-path={target_inspect}"),
              &format!("--port={CONNECTOR_INIT_PORT}"),
        ], args].concat())
        .stderr(Stdio::inherit())
        .spawn()
        .context("spawning docker run child")?;

    Ok((child, dir, port))
}

async fn connector_client(port: u16) -> anyhow::Result<ConnectorClient<tonic::transport::Channel>> {
    loop {
        match ConnectorClient::connect(format!("tcp://127.0.0.1:{port}")).await {
            Ok(client) => return Ok(client),
            Err(_) => {
                std::thread::sleep(std::time::Duration::from_millis(3000));
                continue;
            }
        };
    }
}

pub async fn docker_run(image: &str, req: Request) -> anyhow::Result<Response> {
    let (_child, _dir, port) = docker_spawn(image, &[])?;

    let mut client = connector_client(port).await?;

    let mut response_stream = client.capture(stream::once(async { req })).await?;

    let response = response_stream.get_mut().message().await?;

    return response.ok_or(anyhow!("no response message"))
}

pub async fn docker_run_stream(image: &str, stream: Pin<Box<dyn Stream<Item = Request> + Send + Sync>>) -> anyhow::Result<Pin<Box<dyn TryStream<Item = anyhow::Result<Response>, Ok = Response, Error = anyhow::Error>>>> {
    let (_child, _dir, port) = docker_spawn(image, &[])?;
    let mut client = connector_client(port).await?;
    let response_stream = client.capture(stream).await?;

    Ok(Box::pin(stream::try_unfold(response_stream, |mut rs| async move {
        if let Some(msg) = rs.get_mut().message().await? {
            Ok(Some((msg, rs)))
        } else {
            Ok(None)
        }
    })))
}
