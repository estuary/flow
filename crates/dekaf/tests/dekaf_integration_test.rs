use anyhow::Context;
use dekaf::connector::DekafConfig;
use futures::StreamExt;
use locate_bin;
use rand::Rng;
use rdkafka::{consumer::Consumer, Message};
use schema_registry_converter::async_impl::avro;
use schema_registry_converter::async_impl::schema_registry;
use serde_json::json;
use std::{collections::HashMap, env, io::Write, time::Duration};

async fn create_consumer<'a>(
    username: &'a str,
    token: &'a str,
    topic: &'a str,
) -> (rdkafka::consumer::StreamConsumer, avro::AvroDecoder<'a>) {
    #[allow(non_snake_case)]
    let DEKAF_BROKER = env::var("DEKAF_BROKER").expect("Missing DEKAF_BROKER environment variable");
    #[allow(non_snake_case)]
    let SCHEMA_REGISTRY =
        env::var("DEKAF_REGISTRY").expect("Missing DEKAF_REGISTRY environment variable");

    let consumer: rdkafka::consumer::StreamConsumer = rdkafka::ClientConfig::new()
        .set("bootstrap.servers", DEKAF_BROKER)
        .set("security.protocol", "SASL_PLAINTEXT")
        .set("sasl.mechanism", "PLAIN")
        .set("sasl.username", username)
        .set("sasl.password", token)
        .set("group.id", "this_needs_to_be_set_but_we_dont_use_it")
        .set("enable.auto.commit", "false")
        .set("auto.offset.reset", "smallest")
        .set("enable.auto.offset.store", "false")
        .create()
        .expect("Consumer creation failed");

    consumer
        .subscribe(vec![topic].as_slice())
        .expect("Consumer subscription failed");

    let decoder = avro::AvroDecoder::new(
        schema_registry::SrSettings::new_builder(String::from(SCHEMA_REGISTRY))
            .set_basic_authorization(username, Some(token))
            .build()
            .expect("failed to build avro decoder"),
    );

    (consumer, decoder)
}

#[derive(Debug)]
enum SpecAction {
    Create,
    Delete,
}

async fn test_specs(
    name: &str,
    action: SpecAction,
    mut capture: models::CaptureDef,
    collections: HashMap<&str, models::CollectionDef>,
    mut materialization: models::MaterializationDef,
) -> anyhow::Result<(String, String)> {
    let mut temp_flow = tempfile::NamedTempFile::new()?;

    let suffix: String = rand::thread_rng()
        .sample_iter(&rand::distributions::Alphanumeric)
        .take(4)
        .map(char::from)
        .collect();

    let capture_name = format!("{}/{suffix}/source-http-ingest", name);
    let materialization_name = format!("{}/{suffix}/test-dekaf", name);

    tracing::info!(temp=?temp_flow, "Attempting to {:?}", action);

    // rewrite capture bindings
    capture.bindings.iter_mut().for_each(|binding| {
        binding.target = models::Collection::new(format!("{name}/{}", binding.target.to_string()))
    });

    // rewrite materialization sources
    materialization.bindings.iter_mut().for_each(|binding| {
        binding
            .source
            .set_collection(models::Collection::new(format!(
                "{name}/{}",
                binding.source.collection()
            )))
    });

    let collections_mapped = collections
        .into_iter()
        .fold(HashMap::new(), |mut state, (k, v)| {
            state.insert(format!("{name}/{k}"), v);
            state
        });

    let file_contents = json!({
        "captures": {
            &capture_name: capture
        },
        "collections": collections_mapped,
        "materializations": {
            &materialization_name: materialization
        }
    });

    temp_flow.write(serde_json::to_vec(&file_contents)?.as_slice())?;

    let flowctl = locate_bin::locate("flowctl").context("failed to locate flowctl")?;

    let async_process::Output {
        stderr,
        stdout: _stdout,
        status,
    } = async_process::output(
        async_process::Command::new(flowctl).args(
            match action {
                SpecAction::Create => vec![
                    "catalog",
                    "publish",
                    "--auto-approve",
                    "--default-data-plane",
                    "ops/dp/public/local-cluster",
                    "--source",
                    temp_flow.path().to_str().unwrap(),
                ],
                SpecAction::Delete => vec![
                    "catalog",
                    "delete",
                    "--prefix",
                    name,
                    "--captures=true",
                    "--collections=true",
                    "--materializations=true",
                    "--dangerous-auto-approve",
                ],
            }
            .as_slice(),
        ),
    )
    .await
    .context("failed to invoke flowctl")?;

    if !status.success() {
        let output = String::from_utf8_lossy(&stderr);
        if !output.contains("no specs found matching given selector") {
            anyhow::bail!("flowctl failed: {}", output);
        }
    }

    tracing::info!(
        ?capture_name,
        ?materialization_name,
        "Successful {:?}",
        action
    );

    Ok((capture_name, materialization_name))
}

async fn sops_encrypt(input: models::RawValue) -> anyhow::Result<models::RawValue> {
    #[allow(non_snake_case)]
    let KEYRING = env::var("SOPS_KEYRING").unwrap_or(
        "projects/estuary-control/locations/us-central1/keyRings/sops/cryptoKeys/cd-github-control"
            .to_string(),
    );

    let sops = locate_bin::locate("sops").context("failed to locate sops")?;

    let async_process::Output {
        stderr,
        stdout,
        status,
    } = async_process::input_output(
        async_process::Command::new(sops).args([
            "--encrypt",
            "--input-type",
            "json",
            "--output-type",
            "json",
            "--gcp-kms",
            &KEYRING,
            "--encrypted-suffix",
            "_sops",
            "/dev/stdin",
        ]),
        input.get().as_bytes(),
    )
    .await
    .context("failed to run sops")?;

    if !status.success() {
        anyhow::bail!(
            "decrypting sops document failed: {}",
            String::from_utf8_lossy(&stderr),
        );
    }

    Ok(models::RawValue::from_string(
        std::str::from_utf8(stdout.as_slice())
            .context("failed to parse sops output")?
            .to_string(),
    )?)
}

async fn get_shard_info(
    task_name: &str,
) -> anyhow::Result<proto_gazette::consumer::list_response::Shard> {
    let flowctl = locate_bin::locate("flowctl").context("failed to locate flowctl")?;

    let async_process::Output {
        stderr,
        stdout,
        status,
    } = async_process::output(async_process::Command::new(flowctl).args([
        "raw",
        "list-shards",
        "--task",
        task_name,
        "-ojson",
    ]))
    .await
    .context("failed to list shards")?;

    if !status.success() {
        anyhow::bail!(
            "listing shards failed: {}",
            String::from_utf8_lossy(&stderr),
        );
    }

    Ok(serde_json::from_slice(&stdout)?)
}

async fn wait_for_primary(task_name: &str) -> anyhow::Result<()> {
    loop {
        match get_shard_info(task_name).await {
            Err(e) => {
                tracing::warn!(?e, "Error getting shard info");
                tokio::time::sleep(Duration::from_secs(1)).await;
                continue;
            }
            Ok(info)
                if info.status.iter().any(|status| {
                    status.code() == gazette::consumer::replica_status::Code::Primary
                }) =>
            {
                return Ok(())
            }
            Ok(info)
                if info.status.iter().any(|status| {
                    status.code() == gazette::consumer::replica_status::Code::Failed
                }) =>
            {
                tracing::warn!(statuses = ?info.status, "Shard failed");
                anyhow::bail!("Shard failed");
            }
            Ok(info) => {
                tracing::info!(statuses = ?info.status,"Waiting for primary");
                tokio::time::sleep(Duration::from_secs(1)).await;
                continue;
            }
        }
    }
}

async fn send_docs(task_name: &str, path: &str, docs: Vec<models::RawValue>) -> anyhow::Result<()> {
    let shard = get_shard_info(task_name).await?;

    let shard_endpoint = shard
        .route
        .context("missing shard route")?
        .endpoints
        .first()
        .context("missing shard endpoint")?
        .replace("https://", "");
    let shard_labels = shard
        .spec
        .context("missing shard spec")?
        .labels
        .context("missing shard labels")?
        .labels;

    let hostname = &shard_labels
        .iter()
        .find(|lab| lab.name == labels::HOSTNAME)
        .context("missing HOSTNAME label")?
        .value;
    let port = &shard_labels
        .iter()
        .find(|lab| lab.name == labels::EXPOSE_PORT)
        .context("missing HOSTNAME label")?
        .value;

    let client = reqwest::Client::builder()
        .danger_accept_invalid_certs(true)
        .build()?;
    let url = format!("https://{hostname}-{port}.{shard_endpoint}/{path}");

    tracing::info!(url, "Sending docs");

    for doc in docs {
        let response = client
            .post(url.clone())
            .header("Content-Type", "application/json")
            .body(doc.get().to_string())
            .send()
            .await
            .context("failed to send document")?;

        if !response.status().is_success() {
            let status = response.status();
            let body = response.text().await.unwrap_or_default();
            anyhow::bail!("failed to send document: status={}, body={}", status, body);
        }
    }

    Ok(())
}

// Note: For the moment, this is going to use an externally-running `agent`, either from Tilt
// or the one running in prod. That means that any changes to `dekaf::connector` will not get
// tested, as that functionality gets baked into the agent/runtime directly. An improvement
// to these tests would be to additionally run/test against a local build of the latest `agent`.
async fn roundtrip(
    name: String,
    endpoint_config: DekafConfig,
    schema: serde_json::Value,
    field_selection: serde_json::Value,
    docs: Vec<serde_json::Value>,
) -> anyhow::Result<()> {
    // Create test collection with specific schema
    // Create test Dekaf materialization
    let materialization_config = sops_encrypt(models::RawValue::from_value(&serde_json::to_value(
        endpoint_config.clone(),
    )?))
    .await?;

    let capture: models::CaptureDef = serde_json::from_value(json!({
        "endpoint": {
            "connector": {
                "image": "ghcr.io/estuary/source-http-ingest:dev",
                "config": sops_encrypt(models::RawValue::from_value(&json!({
                    "paths": ["/data"]
                }))).await?
            }
        },
        "bindings": [{
            "resource": {
                "path": "/data",
                "stream": "/data"
            },
            "target": "single_collection"
        }]
    }))?;

    let collections = HashMap::from_iter(
        vec![("single_collection", serde_json::from_value(schema)?)].into_iter(),
    );

    let materialization: models::MaterializationDef = serde_json::from_value(json!({
        "endpoint": {
            "dekaf": {
                "variant": "foo",
                "config": materialization_config
            }
        },
        "bindings": [
            {
                "resource": {
                    "topic_name": "test_topic"
                },
                "source": "single_collection",
                "fields": field_selection
            }
        ]
    }))?;

    let task_name_prefix = format!("test/dekaf_testing/{name}");

    test_specs(
        &task_name_prefix,
        SpecAction::Delete,
        capture.clone(),
        collections.clone(),
        materialization.clone(),
    )
    .await?;

    let (capture_name, materialization_name) = test_specs(
        &task_name_prefix,
        SpecAction::Create,
        capture.clone(),
        collections.clone(),
        materialization.clone(),
    )
    .await?;

    wait_for_primary(&capture_name).await?;

    tracing::info!("Capture is primary");

    send_docs(
        &capture_name,
        "data",
        docs.iter()
            .map(models::RawValue::from_value)
            .collect::<Vec<_>>(),
    )
    .await?;

    tracing::info!("Sent test docs");

    // Consume test documents
    let (consumer, decoder) =
        create_consumer(&materialization_name, &endpoint_config.token, "test_topic").await;

    let mut doc_stream = consumer.stream();

    let mut counter = 0;
    while let Some(consumed) = doc_stream.next().await {
        let consumed = consumed?;

        // Connfirm that field selection was applied

        let decoded_key = match decoder.decode(consumed.key()).await {
            Err(e) => {
                tracing::error!(err=?e, "Error decoding key");
                return Err(anyhow::Error::from(e));
            }
            Ok(d) => apache_avro::from_value::<serde_json::Value>(&d.value),
        }?;

        insta::assert_json_snapshot!(format!("{name}-key-{counter}"), &decoded_key);

        let decoded_payload = match decoder.decode(consumed.payload()).await {
            Err(e) => {
                tracing::error!(err=?e, "Error decoding value");
                return Err(anyhow::Error::from(e));
            }
            Ok(d) => apache_avro::from_value::<serde_json::Value>(&d.value),
        }?;

        insta::assert_json_snapshot!(format!("{name}-value-{counter}"), &decoded_payload, {
            ".flow_published_at.json" => "[timestamp]",
            ".flow_document._flow_extra._meta.json" => "[contains_timestamp]"
        });

        counter += 1;
        if counter >= docs.len() {
            break;
        }
    }

    // Delete test specs

    test_specs(
        &task_name_prefix,
        SpecAction::Delete,
        capture.clone(),
        collections.clone(),
        materialization.clone(),
    )
    .await?;
    Ok(())
}

#[ignore]
#[tokio::test]
#[tracing_test::traced_test]
async fn test_field_selection_specific() -> anyhow::Result<()> {
    roundtrip(
        "field_selection_specific".to_string(),
        dekaf::connector::DekafConfig {
            deletions: dekaf::connector::DeletionMode::Kafka,
            strict_topic_names: false,
            token: "1234".to_string(),
        },
        json!({
            "schema": {
                "properties": {
                    "key": {
                        "type": "string"
                    },
                    "field_a": {
                        "type": "string",
                    },
                    "field_b": {
                        "type": "string",
                    }
                },
                "type": "object",
                "required": [
                    "key",
                    "field_a",
                    "field_b"
                ],
            },
            "key": [
                "/key"
            ]
        }),
        json!({
            "include": {
                "field_a": {}
            },
            "recommended": false
        }),
        vec![json!({
            "key": "first",
            "field_a": "foo",
            "field_b": "bar"
        })],
    )
    .await
}

#[ignore]
#[tokio::test]
#[tracing_test::traced_test]
async fn test_field_selection_recommended() -> anyhow::Result<()> {
    roundtrip(
        "field_selection_recommended".to_string(),
        dekaf::connector::DekafConfig {
            deletions: dekaf::connector::DeletionMode::Kafka,
            strict_topic_names: false,
            token: "1234".to_string(),
        },
        json!({
            "schema": {
                "properties": {
                    "key": {
                        "type": "string"
                    },
                    "field_a": {
                        "type": "string",
                    },
                    "field_b": {
                        "type": "string",
                    }
                },
                "type": "object",
                "required": [
                    "key",
                    "field_a",
                    "field_b"
                ],
            },
            "key": [
                "/key"
            ]
        }),
        json!({
            "recommended": true
        }),
        vec![json!({
            "key": "first",
            "field_a": "foo",
            "field_b": "bar"
        })],
    )
    .await
}

#[ignore]
#[tokio::test]
#[tracing_test::traced_test]
async fn test_field_selection_flow_document() -> anyhow::Result<()> {
    roundtrip(
        "field_selection_flow_document".to_string(),
        dekaf::connector::DekafConfig {
            deletions: dekaf::connector::DeletionMode::Kafka,
            strict_topic_names: false,
            token: "1234".to_string(),
        },
        json!({
            "schema": {
                "properties": {
                    "key": {
                        "type": "string"
                    },
                    "field_a": {
                        "type": "string",
                    },
                    "field_b": {
                        "type": "string",
                    }
                },
                "type": "object",
                "required": [
                    "key",
                    "field_a",
                    "field_b"
                ],
            },
            "key": [
                "/key"
            ]
        }),
        json!({
            "include": {
                "key": {},
                "flow_document": {}
            },
            "exclude": [
                "field_a",
                "field_b"
            ],
            "recommended": true
        }),
        vec![json!({
            "key": "first",
            "field_a": "foo",
            "field_b": "bar"
        })],
    )
    .await
}

#[ignore]
#[tokio::test]
#[tracing_test::traced_test]
async fn test_meta_is_deleted() -> anyhow::Result<()> {
    roundtrip(
        "meta_is_deleted".to_string(),
        dekaf::connector::DekafConfig {
            deletions: dekaf::connector::DeletionMode::CDC,
            strict_topic_names: false,
            token: "1234".to_string(),
        },
        json!({
            "schema": {
                "properties": {
                    "key": {
                        "type": "string"
                    },
                    "field_a": {
                        "type": "string",
                    },
                    "field_b": {
                        "type": "string",
                    },
                    "_meta": {
                        "type": "object",
                        "properties": {
                            "op": {
                                "type": "string"
                            }
                        }
                    }
                },
                "type": "object",
                "required": [
                    "key",
                ],
            },
            "key": [
                "/key"
            ]
        }),
        json!({
            "include": {
                "key": {},
            },
            "recommended": false
        }),
        vec![
            json!({
                "key": "first",
                "field_a": "foo",
                "field_b": "bar",
                "_meta": {
                    "op": "c"
                }
            }),
            json!({
                "key": "first",
                "field_a": "foo",
                "field_b": "bar",
                "_meta": {
                    "op": "d"
                }
            }),
        ],
    )
    .await
}

#[ignore]
#[tokio::test]
#[tracing_test::traced_test]
async fn test_fields_not_required() -> anyhow::Result<()> {
    roundtrip(
        "fields_not_required".to_string(),
        dekaf::connector::DekafConfig {
            deletions: dekaf::connector::DeletionMode::Kafka,
            strict_topic_names: false,
            token: "1234".to_string(),
        },
        json!({
            "schema": {
                "properties": {
                    "key": {
                        "type": "string"
                    },
                    "field_a": {
                        "type": "string",
                    },
                },
                "type": "object",
                "required": [
                    "key",
                ],
            },
            "key": [
                "/key"
            ]
        }),
        json!({
            "recommended": true
        }),
        vec![json!({
            "key": "first",
            // Omitting "field_a"
        })],
    )
    .await
}
