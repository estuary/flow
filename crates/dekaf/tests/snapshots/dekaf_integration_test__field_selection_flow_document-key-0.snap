---
source: crates/dekaf/tests/dekaf_integration_test.rs
assertion_line: 435
expression: "&decoded_key"
---
{
  "_flow_key": {
    "p1": "first"
  }
}
